Python: Fix MCP tool result serialization for list[TextContent]

claude89757 opened the pull request:
## Summary

Fixes #2509

This PR addresses a bug where MCP tool results containing `list[TextContent]` were incorrectly serialized to object repr strings instead of actual text content.

## Problem

When using AG-UI protocol with MCP tools, tool results that returned `list[TextContent]` were being serialized as:
```
[<agent_framework._types.TextContent object at 0x7f8e7c1cc730>]
```

This occurred because the serialization logic in `_events.py` and `_message_adapters.py` only handled `dict` types and fell back to `str()` for all other types, which produced the object repr for lists.

## Solution

Added proper handling for `list` types in both files:

1. Check if `content.result` is a `list`
2. For each item in the list:
   - If it has a `text` attribute (like `TextContent`), extract the text
   - If it has a `model_dump()` method, serialize it to JSON
   - Otherwise, fall back to `str()`
3. If there's only one text item, return it as plain text
4. If there are multiple items, return them as a JSON array

## Changes

- **python/packages/ag-ui/agent_framework_ag_ui/_events.py**: Updated `_handle_function_result_content()` method
- **python/packages/ag-ui/agent_framework_ag_ui/_message_adapters.py**: Updated message adapter logic

## Testing

This fix ensures that MCP tool results are properly displayed in AG-UI frontends like CopilotKit.

## Related Issues

- Fixes #2509

moonbox3 commented on python/packages/ag-ui/agent_framework_ag_ui/_message_adapters.py:
```diff
@@ -242,6 +242,19 @@ def agent_framework_messages_to_agui(messages: list[ChatMessage] | list[dict[str
+                    # Handle Contents list (e.g., from MCP tool results)
+                    import json
+
+                    texts = []
```
This will need to be typed - should it be `text: list[str] = []`?

moonbox3 commented on python/packages/ag-ui/agent_framework_ag_ui/_events.py:
```diff
@@ -386,6 +386,17 @@ def _handle_function_result_content(self, content: FunctionResultContent) -> lis
         result_message_id = generate_event_id()
         if isinstance(content.result, dict):
             result_content = json.dumps(content.result)  # type: ignore[arg-type]
+        elif isinstance(content.result, list):
```
The exact same logic is duplicated in both files. Should this be extracted into a shared utility function (perhaps in `_utils.py` or `_message_adapters.py`) to make sure we have consistency and reduce the maintenance burden? If the serialization logic needs to change, it would need to be updated in both places.

moonbox3 commented on python/packages/ag-ui/agent_framework_ag_ui/_message_adapters.py:
```diff
@@ -242,6 +242,19 @@ def agent_framework_messages_to_agui(messages: list[ChatMessage] | list[dict[str
                     content_text = json.dumps(content.result)  # type: ignore
+                elif isinstance(content.result, list):
+                    # Handle Contents list (e.g., from MCP tool results)
+                    import json
```
we should move the `import json` to the top of the file - it appears twice now (once on line 242 and once here).

claude89757 commented:
Hi @moonbox3, thank you for the detailed review feedback! 🙏

I've addressed all the review comments in the latest commit:

1. ✅ Added `isinstance(item.text, str)` check to ensure the text attribute is actually a string
2. ✅ Fixed the double-serialization issue by keeping `model_dump()` results as dicts until the final `json.dumps()`
3. ✅ Updated the docstring with detailed return value documentation
4. ✅ Added tests for the `_events.py` path covering list type tool results

Please re-review when you have a chance.

*Note: I'm using a "vibe coding" approach with AI assistance, so there might be some edge cases I haven't fully considered. Please feel free to point out anything that looks off — I really appreciate your patience and guidance!*

moonbox3 reviewed (approved):
LGTM, thanks for working through this.

claude89757 commented:
Hi @eavanvalkenburg @moonbox3, I just pushed a small fix for two failing tests to align with `prepare_function_call_results` behavior:

- `test_tool_result_with_none`: Updated expected value to `"null"` (JSON serialization)
- `test_tool_result_with_model_dump_objects`: Changed `MockModel` to use Pydantic `BaseModel`

Could you please re-approve when you have a moment? Thanks!

eavanvalkenburg commented:
@claude89757 several cicd jobs failed, please have a look

claude89757 commented:
Hi @eavanvalkenburg, I've fixed the CI/CD failures. 

The issue was that the `MockTextContent` test helper classes were triggering ruff's B903 linter error ("class could be dataclass or namedtuple"). I've converted them to dataclasses to satisfy the linter check.

All tests and pre-commit checks are now passing locally. Please re-run the CI/CD workflows when you have a chance. Thanks! 🙏
